docs/collector.smbclient.md

```diff
@@ -0,0 +1,37 @@
+# smb collector
+
+The smbclient collector collects metrics from MS SmbClient hosts through perflib
+=======
+
+
+|||
+-|-
+Metric name prefix  | `smbclient`
+Classes 			| [Win32_PerfRawData_SMB](https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-smb/)<br/> 
+Enabled by default? | No
+
+## Flags
+
+### `--collectors.smbclient.list`
+Lists the Perflib Objects that are queried for data along with the perlfib object id
+
+### `--collectors.smbclient.enabled`
+Comma-separated list of collectors to use, for example: `--collectors.smbclient.enabled=ServerShares`. Matching is case-sensitive. Depending on the smb protocol version not all performance counters may be available. Use `--collectors.smbclient.list` to obtain a list of supported collectors.
+
+## Metrics
```

Comment:
I have the feeling that this feature list is not complete.

Reply:
The following table represents what I have in the code at this point. The metrics marked TBD have yet to be verified, or I don't understand the data I'm seeing.  I'm interested in the queue metrics, but as I said having a hard time figuring out what the data is showing me.

Name | Description | Type | Labels
-----|-------------|------|-------
`requests_queued_total` | The number of requests queued on this share. | TBD | `server`, `share` |
`reads_queued` | The number of read requests queued on this share. | TBD | `server`, `share` |
`writes_queued` | The number of write requests queued on this share. | TBD | `server`, `share` |
`request_seconds_total` | Seconds waiting for requests on this share. | counter | `server`, `share` |
`stalls_total` | The number of requests delayed based on insufficient credits on this share. | TBD | `server`, `share` |
`current_data_queues` | The point in time number of requests outstanding on this share. | TBD | `server`, `share` |
`data_bytes_total` | The bytes read or written on this share. | counter | `server`, `share` |
`requests_total` | The requests on this share. | counter | `server`, `share` |
`metadata_requests_total` | The metadata requests on this share. | counter | `server`, `share` |
`read_bytes_via_smbdirect_total` | The bytes read from this share via RDMA direct placement. | TBD | `server`, `share` |
`read_bytes_total` | The bytes read on this share. | counter | `server`, `share` |
`read_requests_via_smbdirect_total` | The read requests on this share via RDMA direct placement. | TBD | `server`, `share` |
`read_requests_total` | The read requests on this share. | counter | `server`, `share` |
`turbo_io_reads_total` | The read requests that go through Turbo I/O | TBD | `server`, `share` |
`turbo_io_writes_total` | The write requests that go through Turbo I/O | TBD | `server`, `share` |
`write_bytes_via_smbdirect_total` | The written bytes to this share via RDMA direct placement. | TBD | `server`, `share` |
`write_bytes_total` | The bytes written on this share. | counter | `server`, `share` |
`write_requests_via_smbdirect_total` | The write requests to this share via RDMA direct placement. | TBD | `server`, `share` |
`write_requests_total` | The write requests on this share. | counter | `server`, `share` |
`read_seconds_total` | Seconds waiting for read requests on this share. | counter | `server`, `share` |
`write_seconds_total` | Seconds waiting for write requests on this share. | counter | `server`, `share` |

Reply:
I will add `windows_smbclient` prefix.

> Based on the source code, everything is marked as counter. Please validate this.

By source code, do you mean `smbclient.go` source? 
Not every metric I added ends with `_total` suffix. For example `writes_queued` where perflib counter name is `perflib:"Avg. Write Queue Length"`. Because it says `Avg.`, does that make it a counter? 

There are 2 situations:
1. The metrics are zero and I cannot tell if they are incrementing. Many of those are marked TBD.
2. I can see the metrics increment and they don't make sense to me. Also marked TBD.

In case 1, we can guess or we can leave them out.
In case 2, I can mark them as counter, but the description will only be a guess and could be wrong. Is it possible to ask MSFT? Has anyone from perf team at MSFT contributed here?

And yes, `current_data_queues`,  `perflib:"Current Data Queue Length"`, is indeed a gauge. I just verified. The value aligns with the number of outstanding I/O 's in our workload generators so it appears to be requests queued. BTW, the metric name doesn't seem right. I will change it to `current_data_queue_length` to be consistent with perflib name. 

The remaining "queue" counters I cannot figure out (for example `perflib:"Avg. Data Queue Length"` I named `requests_queued_total`). I can see counter incrementing. The rate per second increase is in the millions. Way too high to make sense. Is the value bytes, or milliseconds? There's no clue in the name so I don't know.

Reply:
Hi Bob, on https://barreto.home.blog/2012/11/19/windows-server-2012-file-server-tip-new-per-share-smb-client-performance-counters-provide-great-insight/ there is a table comparing smb counters with their physical disk counterparts. Perhaps that helps explaining what they are.

Avg. Data Queue Length in perfmon is described as "The average number of both read and write requests that were queued for this share.", so not the number of bytes, but number of requests. But, millions per second indeed is odd.

Reply:
Also, the exporter converts `Avg. sec/Data Request` from ticks to seconds before it is exposed. The `Avg. Data Queue Length` is exposed raw so the conversion to seconds already has happened.

Reply:
I checked to see how the physical disk collector handles the queue len counters. It simply doesn't expose them. The only queue counter exposed is:
```
type PhysicalDisk struct {
	Name                   string
	CurrentDiskQueueLength float64 `perflib:"Current Disk Queue Length"`
```
This is the same gauge that we already verified.

Reply:
I may have figured it out. I think the value provided by `perflib: "Avg. Data Queue Length"` unit is milliseconds of wait time, represented as ticks (1/10,000). So to convert to seconds, as required by the exporter, we multiply the value by `perflib.TicksToSecondScaleFactor` to convert from ticks to milliseconds, and then multiply by 1/1000 (or millisecond scale factor) to convert from milliseconds to seconds.

We then use wait seconds from above to calculate queue length using Little's Law:
```
  Q = R * W, where,
  Q is the queue len,
  R is the request rate,
  W is wait seconds from above
```
I checked this against performance tests using 5 workloads where we know the number of outstanding IOs in advance (labeled the expected value in table). These were the results:
workload| expected | actual |
-|-|-
prefill seq write | 32 | 31.7 |
rand read | 32 | 28.7 |
seq read | 16 | 14.9 |
seq write | 8 | 7.87 |
rand write | 32 | 31.3 |

Reply:
Also, completely independently, on an entirely different test bed, our SMB performance analyst (not me) confirmed that `perfmon` shows the `Avg. Data Queue Length` equal to 31.82 on `rand write` workload when the expected value is 32. So this is evidence that my queue length calculations align with `perfmon`.

Reply:
Documenting the Prometheus expression I currently use to get the queue length. Important to note that the metric is the raw value returned from `perflib:"Avg. Data Queue Length"`, the exporter metric name is tentative and could change.
```
irate(windows_smbclient_requests_queued_total[4s]) * (1/10000000)
```
The confusing part is that when consider the counter as ticks (time) and convert to seconds, it produces the queue length (not a time value).